src/ctia/entity/event/obj_to_event.clj

```diff
@@ -41,27 +41,146 @@
             :action "added"
             :change {}})
          (remove #(contains? diff-before %)
-                 (keys diff-after))))))
+                 (keys diff-after)))))
 
+(defn truncate
+  "Truncate v, a possibly-nested value, by inserting placeholder after given
+  max-depth and max-count."
+  ([v placeholder max-count max-depth] (truncate v placeholder max-count max-depth 0))
+  ([v placeholder max-count max-depth depth]
+   (let [tru #(truncate % placeholder max-count max-depth (inc depth))
+         dots placeholder]
+     (cond
+       (< max-depth depth) dots
+       (seq? v) (let [[pre suf] (split-at max-count v)]
+                  (doall
+                    (concat (map tru pre)
+                            (when (seq suf)
+                              [dots]))))
+       (set? v) (let [pre (take max-count v)
+                      ;; in case keys conflict after truncation,
+                      ;; remove all conflicting keys.
+                      ;; eg., if kvs is [:a :a :b]
+                      ;;      then truncate to [:b]
+                      ks (mapv tru pre)
+                      s (->> ks
+                             frequencies
+                             (filter (comp #{1} second))
+                             (map first)
+                             set)]
+                  (cond-> s
+                    (not= (count s) (count v)) (conj dots)))
+
+       (map? v) (let [pre (take max-count (keys v))
+                      ;; in case keys conflict after truncation,
+                      ;; remove all conflicting keys.
+                      ;; eg., if kvs is [[:a 1] [:a 2] [:b 3]]
+                      ;;      then truncate to [[:b 3]]
```

Comment:
why truncating keys?

Reply:
I'm treating each time we recur into a collection as 1 depth, so if the depth limit is 1 we truncate `{:a 1 :b 2 :c 2}` to `{... ...}` instead of `{:a ... :b ... :c ...}`.

Reply:
Ok I have not understood it properly.